오픈마켓 [STEP2] 아샌, 제인

siwonkim0 opened the pull request:
안녕하세요. 코리! @corykim0829

아샌(@ICS-Asan), 제인(@siwonkim0) 입니다. 

CollectionView에 대해 공부하고 구현하는 것이 어려워 기한보다 조금 늦게 PR올렸습니다.

많이 부족하겠지만 이번 PR도 잘 부탁드리겠습니다!!

감사합니다!!

### 💻  주요 구현 내용

### < ViewController >

- 화면 전환시 View를 새로 생성하지 않고 View를 미리 생성해놓고 바꿔주는 형식으로 구현
    - `listCollectionView`, `gridCollectionView` 두가지 뷰를 생성
    - 각 CollectionView에 맞게 CollectionViewDiffableDataSource생성 및 세팅하는 메서드 구현
- 첫 화면은 ListView가 나오도록 구현
- segementControl을 통해 알맞는 CollectionView를 View에 할당해줌
    - 화면전환시 새로 View를 생성하지 않기 때문에 빠르게 전환이 가능
- `OpenMarketViewLayout` 구조체 내부에 static let으로 각 layout을 선언하여 CollectionView생성시 사용할 수 있도록 구현
    - grid화면의 경우 각 cell의 크기는 고정해두어야 일정한 화면 구성이 될 것이라 판단하여 cell 크기 고정

### < CustomCollectionViewCell >

- defaultContentConfiguration()를 사용하지 않고 원하는 View를 커스텀하기 위해 List와 Grid에서 사용할 Cell을 구현
- 각 Cell의 xib파일에 AutoLayout과 각 요소의 기본값을 설정
- Label에서 자주 사용되는 특정 문자나 띄워쓰기, 줄바꿈 등을 관리하기 위한 NameSpace를 생성하여 사용
- `String+extension`, `Int+extension`을 활용하여 반복적으로 사용되는 로직을 편리하게 사용할 수 있도록 구현

### < 🤔고민했던 부분 - ListCollectionView와 GridCollectionView간에 View 전환시 로딩이 느리게 된 부분>

- 수정 전
    
    > collectionView와 dataSource를 각각 하나씩만 두고 layout과 cell을 list/grid 상황에 맞게 바꿨습니다. 
    그러나 `ListCollectionView`와 `GridCollectionView`간에 전환을 시킬때 전환이 되고나서 뷰가 느리게 바뀌는 상황이 발생하였습니다.
    해결방법을 찾다가 공식문서를 보니, `DiffableDataSource` 로 데이터를 한번 configure하고나서 내용을 바꾸지 말라는 주의사항이 있었습니다.
    > 

> **UICollectionViewDiffableDataSource 공식문서**
`Don’t change the dataSource` on the collection view after you configure it with a diffable data source. If the collection view needs a new data source after you configure it initially, create and configure a new collection view and diffable data source.
> 

```swift
let collectionView = UICollectionView(frame: .zero, collectionViewLayout: OpenMarketViewLayout.list)
var dataSource: UICollectionViewDiffableDataSource<Section, ProductDetail>?
var isListLayout = true

func setupCollectionView() {
        dataSource = UICollectionViewDiffableDataSource<Section, ProductDetail>(collectionView: collectionView) { collectionView, indexPath, product in
            if self.isListLayout {
                return self.productListCell(indexPath: indexPath, with: product)
            }
            return self.productGridCell(indexPath: indexPath, with: product)
        }
    }
  
func productListCell(indexPath: IndexPath, with product: ProductDetail) -> UICollectionViewCell {
    guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ProductListCell.identifier, for: indexPath) as? ProductListCell else {
        return UICollectionViewListCell()
    }
    cell.setup(with: product)
    return cell
}

func productGridCell(indexPath: IndexPath, with product: ProductDetail) -> UICollectionViewCell {
    guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ProductGridCell.identifier, for: indexPath) as? ProductGridCell else {
        return UICollectionViewListCell()
    }
    cell.setup(with: product)
    return cell
}
```

```swift
@IBAction func segementChanged(_ sender: UISegmentedControl) {
    switch sender.selectedSegmentIndex {
    case 0:
        isListLayout = true
        collectionView.collectionViewLayout = OpenMarketViewLayout.list
        collectionView.reloadData()
    case 1:
        isListLayout = false
        collectionView.collectionViewLayout = OpenMarketViewLayout.grid
        collectionView.reloadData()
    default:
        return
    }
}
```

- 수정 후

> 공식문서의 안내에 따라 생성된 `DiffableDataSource`의 내용을 변경하지 않고, 새로운 `DiffableDataSource`필요시 collectionView를 생성하도록 구현하였습니다.
collectionView와 dataSource 를 각각 두개씩 두고 뷰 자체를 list/grid 상황에 맞게 바꿨습니다.
> 

```swift
let listCollectionView = UICollectionView(frame: .zero, collectionViewLayout: OpenMarketViewLayout.list)
let gridCollectionView = UICollectionView(frame: .zero, collectionViewLayout: OpenMarketViewLayout.grid)
var listDataSource: UICollectionViewDiffableDataSource<Section, ProductDetail>?
var gridDataSource: UICollectionViewDiffableDataSource<Section, ProductDetail>?

func setupListCollectionView() {
    listDataSource = UICollectionViewDiffableDataSource<Section, ProductDetail>(collectionView: listCollectionView) { collectionView, indexPath, product in
        
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ProductListCell.identifier, for: indexPath) as? ProductListCell else {
            return UICollectionViewListCell()
        }
        cell.setup(with: product)
        return cell
    }
}

func setupGridCollectionView() {
    gridDataSource = UICollectionViewDiffableDataSource<Section, ProductDetail>(collectionView: gridCollectionView) { collectionView, indexPath, product in
        
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ProductGridCell.identifier, for: indexPath) as? ProductGridCell else {
            return UICollectionViewListCell()
        }
        cell.setup(with: product)
        return cell
    }
}
```

```swift
@IBAction func segementChanged(_ sender: UISegmentedControl) {
        switch sender.selectedSegmentIndex {
        case 0:
            view = listCollectionView
        case 1:
            view = gridCollectionView
        default:
            return
        }
    }
```

## ❓   질문드리고 싶은 내용

### < CollectionView에 하나의 DataSource만 사용해야하는지 >

- 위 고민했던부분에서 저희는 modern한 방식인 UICollectionViewDiffableDataSource 를 사용해서 한 UIViewController 안에서 두가지의 DataSource를 같이 사용하였습니다.
- 공식문서에서 `DiffableDataSource`를 사용한 이후에는 컬렉션뷰의 dataSource를 바꾸지 말라고 되어있는데... 그렇다면 `DiffableDataSource`가 아닌 그냥 `DataSource`를 사용하는 이전 방식에서는 하나의 컬렉션뷰에 두개의 데이터소스를 바꿔가며 사용해도 되는지 궁금합니다.
- UICollectionViewDataSource 프로토콜을 채택한 타입 두개를 만들고 각각의 cellForItemAt 메서드에서 셀에 데이터를 세팅해준 후 뷰컨트롤러의 프로퍼티로 설정해주고 뷰컨트롤러의 datasource = 새로 생성한 타입 으로 설정해주면 되지 않을까라고 생각해보았습니다!
- 그런데 이렇게 되면 저희가 겪었던 문제처럼 list컬렉션뷰와 grid컬렉션뷰간에 전환을 시킬때 전환이 되고나서 뷰가 느리게 바뀌는 상황이 발생할까요...? 그렇다면 보통 한 뷰 안에서 컬렉션뷰의 레이아웃을 바꿀 경우 어떤 방식으로 구현을하는게 좋을지 궁금합니다.

```swift
class DataSource: NSObject, UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        //
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        //
    }
}
```

### < ContentConfiguration vs XIB 커스텀 셀 >

- 저희가 CollectionView의 셀을 구현하기 위해 `defaultContentConfiguration()사용 → 직접 구현한 UIContentConfiguration 사용 → CustomCell(with Xib) 사용`의 과정을 거쳐 현재의 코드가 되었습니다.
- 구현하다 보니 현재 사용한 방법이 코드도 간결하고 구현하기도 쉬웠는데 현업에서는 어떠한 방법을 선호하는지 이유는 무엇인지 궁금합니다.


감사합니다 👍

siwonkim0 committed:
feat: ProductListCell 생성 #7

- List collectionView에 들어갈 셀 구현
- 오토레이아웃 설정

ICS-Asan committed:
feat: ProductListCell setup() 구현 #7

- setupCellAccessory()
- setupImage()
- setupPriceLabel()
- setupStockLabel()
- 사용되는 문자열을 관리하는 StringSeparator, LabelString NameSpace생성

siwonkim0 committed:
feat: ProductListView 생성 #7

- custom UICollectionView 생성하여 ViewController의 view에 ProductListView 할당하여 메인뷰가 컬렉션뷰가 되도록함

ICS-Asan committed:
feat: ViewController 내부 구현 #7

- Section 타입 생성
- apiManager, dataSource 프로퍼티 생성
- getProducts, setupCollectionView, populate 메서드 생
- accessory indicator 위치 설정 불가로 인한 이미지뷰 대체

siwonkim0 committed:
feat: ProductListCell 폰트수정, segment control 버튼 추가 #7

ICS-Asan committed:
feat: ProductGridCell 생성 및 구현, 파일 분리 작업 #8

- ProductGridCell과 ProductListCell에서 공용으로 사용되는 부분 파일 분리

siwonkim0 committed:
feat: ProductGridCell 오토레이아웃 적용 #8

ICS-Asan committed:
feat: OpenMarketViewLayout타입 생성 및 구현 #8

- ProductListView 삭제
- ViewController 내부 isListLayout 프로퍼티 생성
- ProductGridCell collectionView에 등록

siwonkim0 committed:
feat: setupCollectionView 구현부 수정 #8

- 레이아웃에 따라 다른 셀이 dequeue되도록 로직 수정

ICS-Asan committed:
feat: segementChanged IBAction 메서드 생성 #8

- segementContorller의 값에 따라 DataSource와 Layout을 변경

siwonkim0 committed:
refactor: 컬렉션뷰 형식 변경시 뷰가 느리게 전환되는 현상 개선 #8

- listCollectionView와 gridCollectionView를 각각 생성해서 view에 할당하여 사용
- listDataSource와 gridDataSource를 각각 생성해서 collectionView의 종류에 따라 다른 dataSource 사용
- 컬렉션뷰 형식 변경시 오토레이아웃 충돌이 발생하는 현상도 해결됨

ICS-Asan committed:
refactor: ViewController네이밍 수정 및 내부 매서드 수정 #8

- ViewController->OpenMarketViewController
- setupSegmentedControl(), registerCollectionViewCell() 생성
- 메서드 호출순서에 따라 메서드 구현 순서 변경

siwonkim0 committed:
refactor: ProductGridCell UI수정 #8

- stockLabel의 textColor가 잘못나오는 부분 수정
- 셀의 크기가 작아지면 imageView가 우선적으로 줄어들도록 오토레이아웃 수정

ICS-Asan committed:
feat: Int Extension 생성 및 내부 addComma() 구현 #8

- 파일 정리

ICS-Asan committed:
refactor: 접근제어 키워드 추가 #8

corykim0829 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    private func registerCollectionViewCell() {
+        listCollectionView.register(UINib(nibName: "ProductListCell", bundle: nil), forCellWithReuseIdentifier: ProductListCell.identifier)
+        gridCollectionView.register(UINib(nibName: "ProductGridCell", bundle: nil), forCellWithReuseIdentifier: ProductGridCell.identifier)
```
nibName도 변수로 만들어서 관리하거나 `String(describing:)` 함수를 이용해서 해당 클래스명을 그대로 사용할 수 있어요!

siwonkim0 replied:
오오 그런 방법이 있었군요..! 피드백 반영하여 수정하였습니다 ~ 😁

```Swift
private func registerCollectionViewCell() {
        listCollectionView.register(UINib(nibName: String(describing: ProductListCell.self), bundle: nil), forCellWithReuseIdentifier: ProductListCell.identifier)
        gridCollectionView.register(UINib(nibName: String(describing: ProductGridCell.self), bundle: nil), forCellWithReuseIdentifier: ProductGridCell.identifier)
    }
```

corykim0829 commented on OpenMarket/OpenMarket/CustomCell/ProductGridCell.swift:
```diff
@@ -0,0 +1,70 @@
+    @IBOutlet weak var gridPriceLabel: UILabel!
+    @IBOutlet weak var gridStockLabel: UILabel!
+    
+    static let identifier = "GridCell"
```
identifier를 static 변수로 처리한 부분 깔끔하고 좋네요 👍

corykim0829 commented on OpenMarket/OpenMarket/CustomCell/ProductGridCell.swift:
```diff
@@ -0,0 +1,70 @@
+        layer.borderColor = UIColor.systemGray.cgColor
+    }
+    
+    func setup(with product: ProductDetail) {
```
외부에서 모델 값만 넘겨주면 뷰단인 Cell에서 업데이트 하는 로직이 깔끔하고 좋습니다 👍

corykim0829 commented on OpenMarket/OpenMarket/Extensions/Int+extension.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+extension Int {
+    func addComma() -> String {
```
편하게 사용할 수 있는 extension 함수 좋아요! 👍

corykim0829 commented on OpenMarket/OpenMarket/Extensions/Int+extension.swift:
```diff
@@ -0,0 +1,10 @@
+    func addComma() -> String {
+        let numberformatter = NumberFormatter()
+        numberformatter.numberStyle = .decimal
+        let convertedNumber = numberformatter.string(for: self) ?? ""
```
nil값이 나와버리면 ""을 반환하게 되는데, 다른 방법은 없을까요?

siwonkim0 replied:
사실.... 현재 Int와 Double의 extension으로 사용되고 있기 때문에 nil이 나올 가능성이 전혀 없다고 판단되는데, 이러한 경우에는 강제 언래핑을 사용해도 될지 궁금합니다.

corykim0829 replied:
강제 언래핑 사용해도 괜찮을 것 같아요!

corykim0829 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewLayout.swift:
```diff
@@ -0,0 +1,21 @@
+        return layout
+    }()
+    
+    static let grid: UICollectionViewFlowLayout = {
```
여기 안에서 매직넘버들을 변수로 만들어서 관리하면 유지보수에 더 좋을 것 같아요!

siwonkim0 replied:
```swift
 private static let gridSectionInset = UIEdgeInsets.init(top: 5, left: 10, bottom: 5, right: 10)
    
 private static let gridItemSize: CGSize = {
    let halfWidth = UIScreen.main.bounds.width / 2.0
    return CGSize(width: halfWidth - 15, height: 260)
}()
    
 private static let gridMinimumInteritemSpacing: CGFloat = 0
static let grid: UICollectionViewFlowLayout = {
    let layout = UICollectionViewFlowLayout()
    layout.sectionInset = gridSectionInset
    layout.minimumInteritemSpacing = gridMinimumInteritemSpacing
    layout.itemSize = gridItemSize
    return layout
}()
```
코멘트주신대로 OpenMarketViewLayout 타입 내에 static let으로 상수를 만들어서 수정해보았습니다. 

매직넘버들만 따로 빼보니 한번만 사용될 매직넘버들을 분리해놓는것이 오히려 비효율인것 같고 타입도 CGFloat라서 타입을 명시까지 해줘야해서 코드가 길어지고 가독성이 떨어진다는 느낌을 받았습니다... 

따라서 매직넘버들만 따로빼지않고 인스턴스들을 따로 상수로 만들어주었는데 혹시 이런 방법을 의도하신걸까요~?

corykim0829 replied:
struct 내부에 다음과 같이 보통 상수를 정리하면 한눈에 보기에도 좋고 유지보수에도 좋습니다!
```Swift
enum Metric {
    static let constant1: CGFloat = 1.2
    static let constant2: CGFloat = 1.2
    static let constant3: CGFloat = 1.2
    static let constant4: CGFloat = 1.2
}

corykim0829 reviewed (changes requested):
수고 정말 많으셨습니다 아샌, 제인 🙌🙌🙌🙌🙌
2주 동안 너무 고생 많으셨습니다.
이번 미션에서 모던 CollectionViewDiffable에 도전해본 점 매우 좋은 것 같습니다 👍
하지만 아직 레거시 코드들에는 UICollectionView을 많이 사용하고 있기 때문에 모두 구현해보시고 차이점을 알아보시는 것도 추천합니다!

제가 남겨드린 코멘트 바탕으로 개선해보면 좋을 것 같아요 :)

---

### CollectionView에 하나의 DataSource만 사용해야하는지

- 위 고민했던부분에서 저희는 modern한 방식인 UICollectionViewDiffableDataSource 를 사용해서 한 UIViewController 안에서 두가지의 DataSource를 같이 사용하였습니다.
- 공식문서에서 DiffableDataSource를 사용한 이후에는 컬렉션뷰의 dataSource를 바꾸지 말라고 되어있는데... 그렇다면 DiffableDataSource가 아닌 그냥 DataSource를 사용하는 이전 방식에서는 하나의 컬렉션뷰에 두개의 데이터소스를 바꿔가며 사용해도 되는지 궁금합니다.
- UICollectionViewDataSource 프로토콜을 채택한 타입 두개를 만들고 각각의 cellForItemAt 메서드에서 셀에 데이터를 세팅해준 후 뷰컨트롤러의 프로퍼티로 설정해주고 뷰컨트롤러의 datasource = 새로 생성한 타입 으로 설정해주면 되지 않을까라고 생각해보았습니다!
- 그런데 이렇게 되면 저희가 겪었던 문제처럼 list컬렉션뷰와 grid컬렉션뷰간에 전환을 시킬때 전환이 되고나서 뷰가 느리게 바뀌는 상황이 발생할까요...? 그렇다면 보통 한 뷰 안에서 컬렉션뷰의 레이아웃을 바꿀 경우 어떤 방식으로 구현을하는게 좋을지 궁금합니다.

➡️ 보통 Datasource는 데이터에 관여하는 역할이기 때문에 레이아웃만 다르고 같은 데이터를 사용할 때에는 하나만 사용해도 됩니다! 
레이아웃 전환 시에 느리게 전환되는 부분은 이와 별개로 뷰단을 업데이트 해줄 때 발생합니다! 네트워크 통신 후에 뷰를 메인 스레드에서 업데이트 해줘야합니다. 또한 이미지 네트워크 통신 시에 사용하는 `Data(contentsOf:)` 때문에 느리게 전환이 되는 것처럼 보이게 되는건데요! 공식문서를 한번 확인해보고 고쳐보면 좋을 것 같습니다!

### ContentConfiguration vs XIB 커스텀 셀

- 저희가 CollectionView의 셀을 구현하기 위해 defaultContentConfiguration()사용 → 직접 구현한 UIContentConfiguration 사용 → CustomCell(with Xib) 사용의 과정을 거쳐 현재의 코드가 되었습니다.
- 구현하다 보니 현재 사용한 방법이 코드도 간결하고 구현하기도 쉬웠는데 현업에서는 어떠한 방법을 선호하는지 이유는 무엇인지 궁금합니다.

➡️ 이거는 팀마다 작업마다 바뀔 수 있는 부분인 것 같습니다. 저 같은 경우에는 xib 또는 모두 코드로 작성하는 것을 선호합니다. xib로 하게 되면, 해당 cell의 뷰를 xib로 관리할 수 있어서 눈으로 직접 UI를 보고 관리할 수 있어서 좋습니다. 코드로 작성하면 xib에 대한 충돌을 줄일 수 있다는 장점이 더 추가되지만 UI를 바로 확인하기 어렵다는 단점이 있죠!

siwonkim0 committed:
refactor: String -> URL -> ImageData 디코딩 과정에서 처리 #9

- ProductDetail의thumbnail 프로퍼티 타입 String -> Data로 변경
- JSONParser에 imageDataDecoder 메서드 생성 및 호출
- ProductListCell에서 불러오던 image를 JSONParser에서 처리
- 추가 버그 리팩토링

ICS-Asan committed:
refactor: registerCollectionViewCell 내부 하드코딩 개선 #9

- register메서드의 nibName 파라미터에 String(describing: )사용

ICS-Asan commented:
알려주신대로 이미지 네트워크 통신 시에 사용하는 `Data(contentsOf:)`를 확인 해봤습니다.
저희가 내린 결론은 뷰의 image를 setup하는 과정에서 해당 이니셜라이저를 사용하지 않고 데이터를 decode하는 과정에서 처리를 하면 뷰에 데이터를 넣을 때마다 네트워킹을 하지 않아도 될 것이라 판단했습니다.
```swift
static func decodeData<T: Decodable>(of JSON: Data, type: T.Type) -> T? {
        let decoder = JSONDecoder()
        
        var decodedData: T?
        do {
            decoder.dataDecodingStrategy = .custom(imageDataDecoder)
            decodedData = try decoder.decode(type, from: JSON)
         ...    
        return decodedData
    }
static func imageDataDecoder(decoder: Decoder) throws -> Data {
        let container = try decoder.singleValueContainer()
        let str = try container.decode(URL.self)
        let imageData = try Data(contentsOf: str)
        
        return imageData
    }
```
그래서 imageDataDecoder라는 메서드를 구현하여 JSON데이터를 decode하는 과정에서 바로 Data로 저장할 수 있도록 구현해봤습니다.

corykim0829 commented:
수고 많으셨습니다! `Data(contentsOf:)` 자체가 동기처리를 하기 때문에 앱이 느려지는 현상이 발생할 수 있는데요! 비동기로 처리하는 방법을 한번 찾아보시고 여유가되면 해보시면 좋을 것 같아요 ㅎㅎ
